reactivesocket-core/src/main/java/io/reactivesocket/internal/SafeCancellableSubscriber.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2016 Netflix, Inc.
+ * <p>
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ *  the License. You may obtain a copy of the License at
+ *  <p>
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *  <p>
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations under the License.
+ */
+
+package io.reactivesocket.internal;
+
+import org.reactivestreams.Subscription;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+abstract class SafeCancellableSubscriber<T> extends CancellableSubscriber<T> {
+
+    protected final AtomicBoolean subscribed = new AtomicBoolean();
+    protected final AtomicBoolean done = new AtomicBoolean();
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        if (subscribed.compareAndSet(false, true)) {
+            super.onSubscribe(s);
+        } else {
+            onError(new IllegalStateException("Duplicate subscription."));
+        }
```

Comment:
Very glad to hear that 👍 
I'll keep my eye on the repo, hopefully being able to provide some useful feedback :)